File: commands/history.js
import { SlashCommandBuilder, EmbedBuilder } from 'discord.js';
import { readTournamentData, readAllVotes, readUserWagers, readCurses } from '../utils/firebase.js';
import { getWinner, getMatchVote } from '../utils/helper.js';
import { withErrorHandler, getTournamentName } from '../utils/command.js';

export const data = new SlashCommandBuilder()
  .setName('history')
  .setDescription('Lật lại lịch sử vote — đúng sai không trốn được')
  .addUserOption(option => option.setName('user')
    .setDescription('Soi ai (bỏ trống = soi mình)')
    .setRequired(false))
  .addIntegerOption(option => option.setName('count')
    .setDescription('Số trận muốn xem (mặc định 5)')
    .setMinValue(1)
    .setMaxValue(20)
    .setRequired(false));

export const execute = withErrorHandler(async (interaction) => {
  await interaction.deferReply();

  const tournamentName = await getTournamentName();
  const targetUser = interaction.options.get('user')?.user || interaction.user;
  const userId = targetUser.id;
  const count = interaction.options.get('count')?.value || 5;

  const allMatches = await readTournamentData('matches');
  if (!allMatches) {
    await interaction.editReply({ content: '❌ Không có dữ liệu trận đấu.' });
    return;
  }

  const [votes, userWagers, allCurses] = await Promise.all([
    readAllVotes(),
    readUserWagers(userId),
    readCurses(),
  ]);
  const users = interaction.client.cachedUsers;
  const nickname = users[userId]?.nickname || targetUser.displayName;

  const completedMatches = allMatches
    .filter((m) => m.hasResult && m.isCalculated)
    .sort((a, b) => Date.parse(a.date) - Date.parse(b.date));

  const userHistory = [];
  const cursedTargets = {};

  for (const match of completedMatches) {
    const key = `${match.id - 1}`;
    const winner = getWinner(match);
    const userVote = getMatchVote(votes, key, match.messageId, userId);

    const wager = userWagers[match.id];
    const curse = allCurses[match.id]?.[userId];

    if (curse) {
      cursedTargets[curse.target] = (cursedTargets[curse.target] || 0) + 1;
    }

    userHistory.push({
      matchId: match.id,
      home: match.home,
      away: match.away,
      result: match.result,
      winner,
      vote: userVote,
      correct: userVote ? userVote === winner : null,
      wagerType: wager?.type || null,
      curseTarget: curse?.target || null,
    });
  }

  const recent = userHistory.slice(-count).reverse();

  if (recent.length === 0) {
    const embed = new EmbedBuilder()
      .setTitle('🔍  Không có lịch sử')
      .setDescription(`Chưa có trận nào xong cho **${nickname}** cả.`)
      .setColor(0xFEE75C);
    await interaction.editReply({ embeds: [embed] });
    return;
  }

  const lines = recent.map((r) => {
    const score = `${r.result.home}-${r.result.away}`;
    let line;
    if (r.vote === null) {
      const autoLabel = r.wagerType === 'random' ? 'ngẫu nhiên 🎲' : 'tự động (least-voted)';
      line = `🤖 **#${r.matchId}** ${r.home.toUpperCase()} ${score} ${r.away.toUpperCase()} — *${autoLabel}*`;
    } else {
      const icon = r.correct ? '👑' : '🤡';
      line = `${icon} **#${r.matchId}** ${r.home.toUpperCase()} ${score} ${r.away.toUpperCase()} — vote **${r.vote.toUpperCase()}**`;
    }

    const tags = [];
    if (r.wagerType === 'double-down') tags.push('⏫ double-down');
    if (r.curseTarget) {
      const targetName = users[r.curseTarget]?.nickname || 'Unknown';
      tags.push(`🪄 nguyền **${targetName}**`);
    }
    if (tags.length > 0) line += `\n  └ ${tags.join(' · ')}`;

    return line;
  });

  const totalVoted = userHistory.filter((r) => r.vote !== null).length;
  const totalCorrect = userHistory.filter((r) => r.correct === true).length;
  const totalRandom = userHistory.filter((r) => r.vote === null && r.wagerType === 'random').length;
  const totalAutoAssigned = userHistory.filter((r) => r.vote === null && r.wagerType !== 'random').length;
  const winRate = totalVoted > 0 ? `${Math.round((totalCorrect / totalVoted) * 100)}%` : '—';

  let summary = `🎯 Tỉ lệ đúng **${winRate}** (${totalCorrect}/${totalVoted})`;
  if (totalRandom > 0) summary += ` · 🎲 ${totalRandom} random`;
  if (totalAutoAssigned > 0) summary += ` · 🤖 ${totalAutoAssigned} tự động`;

  const totalDD = userHistory.filter((r) => r.wagerType === 'double-down').length;
  const totalCurses = userHistory.filter((r) => r.curseTarget).length;
  if (totalDD > 0 || totalCurses > 0) {
    const parts = [];
    if (totalDD > 0) parts.push(`⏫ ${totalDD} double-down`);
    if (totalCurses > 0) parts.push(`🪄 ${totalCurses} curse`);
    summary += `\n${parts.join(' · ')}`;
  }

  if (Object.keys(cursedTargets).length > 0) {
    const curseList = Object.entries(cursedTargets)
      .sort((a, b) => b[1] - a[1])
      .map(([id, count]) => `${users[id]?.nickname || 'Unknown'} (×${count})`)
      .join(', ');
    summary += `\n🪄 Nguyền: ${curseList}`;
  }

  const embed = new EmbedBuilder()
    .setTitle(`📜  Lịch sử vote ${nickname}`)
    .setDescription(
      `**${tournamentName}**\n\n` +
      lines.join('\n') +
      `\n\n${summary}`
    )
    .setColor(0x5865F2)
    .setThumbnail(targetUser.displayAvatarURL())
    .setFooter({ text: `Hiển thị ${recent.length}/${userHistory.length} trận gần nhất` })
    .setTimestamp();

  await interaction.editReply({ embeds: [embed] });
});

File: commands/history.go
package commands

import (
	"context"
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/kdnvi/rambo-bot/bot"
	"github.com/kdnvi/rambo-bot/internal/discord"
)

func History(ctx context.Context, b *bot.Bot, i *discord.Interaction) {
	if err := deferReply(ctx, b, i, true); err != nil {
		return
	}

	tournamentName := getTournamentName(ctx, b)

	targetID := i.ActingUser().ID
	targetDisplayName := i.ActingUser().DisplayName()
	targetAvatarURL := i.ActingUser().AvatarURL()

	if opt := i.Data.GetOption("user"); opt != nil {
		targetID = opt.String()
	}
	count := int64(5)
	if opt := i.Data.GetOption("count"); opt != nil {
		count = opt.Int()
	}

	if u := b.GetCachedUser(targetID); u != nil {
		targetDisplayName = u.DisplayName()
		targetAvatarURL = u.AvatarURL
	}

	matches, err := b.DB.ReadMatches(ctx)
	if err != nil || matches == nil {
		editReply(ctx, b, i, discord.EditMessagePayload{Content: "❌ Không có dữ liệu trận đấu."})
		return
	}

	votes, _ := b.DB.ReadAllVotes(ctx)
	userWagers, _ := b.DB.ReadUserWagers(ctx, targetID)
	allCurses, _ := b.DB.ReadCurses(ctx)
	users := b.GetAllCachedUsers()

	// Completed matches sorted oldest first
	var completed []int
	for idx, m := range matches {
		if m.HasResult && m.IsCalculated {
			completed = append(completed, idx)
		}
	}
	sort.Slice(completed, func(a, b int) bool {
		ta, _ := time.Parse(time.RFC3339, matches[completed[a]].Date)
		tb, _ := time.Parse(time.RFC3339, matches[completed[b]].Date)
		return ta.Before(tb)
	})

	type histEntry struct {
		matchID      int
		home, away   string
		result       [2]int
		winner       string
		vote         string
		correct      *bool
		hasDoubleDown bool
		hasRandom    bool
		curseTarget  string
	}

	cursedTargetCounts := map[string]int{}
	var history []histEntry

	for _, idx := range completed {
		m := matches[idx]
		key := fmt.Sprintf("%d", m.ID-1)
		winner := getWinner(&m)
		userVote := getMatchVote(votes, key, m.MessageID, targetID)

		matchIDStr := fmt.Sprintf("%d", m.ID)
		wager := userWagers[matchIDStr]
		curse := allCurses[fmt.Sprintf("%d", m.ID)]
		var curseTarget string
		if c, ok := curse[targetID]; ok {
			curseTarget = c.Target
			cursedTargetCounts[curseTarget]++
		}

		var correct *bool
		if userVote != "" {
			c := userVote == winner
			correct = &c
		}

		res := [2]int{}
		if m.Result != nil {
			res = [2]int{m.Result.Home, m.Result.Away}
		}

		history = append(history, histEntry{
			matchID:      m.ID,
			home:         m.Home,
			away:         m.Away,
			result:       res,
			winner:       winner,
			vote:         userVote,
			correct:      correct,
			hasDoubleDown: wager != nil && wager.DoubleDown,
			hasRandom:    wager != nil && wager.Random,
			curseTarget:  curseTarget,
		})
	}

	// Take last `count` entries, reverse to newest-first
	recent := history
	if int64(len(recent)) > count {
		recent = recent[len(recent)-int(count):]
	}
	for lo, hi := 0, len(recent)-1; lo < hi; lo, hi = lo+1, hi-1 {
		recent[lo], recent[hi] = recent[hi], recent[lo]
	}

	if len(recent) == 0 {
		editReply(ctx, b, i, discord.EditMessagePayload{
			Embeds: []discord.Embed{{
				Title:       "🔍  Không có lịch sử",
				Description: fmt.Sprintf("Chưa có trận nào xong cho **%s** cả.", targetDisplayName),
				Color:       0xFEE75C,
			}},
		})
		return
	}

	lines := make([]string, 0, len(recent))
	for _, r := range recent {
		score := fmt.Sprintf("%d-%d", r.result[0], r.result[1])
		var line string
		if r.vote == "" {
			autoLabel := "tự động (least-voted)"
			if r.hasRandom {
				autoLabel = "ngẫu nhiên 🎲"
			}
			line = fmt.Sprintf("🤖 **#%d** %s %s %s — *%s*",
				r.matchID, upper(r.home), score, upper(r.away), autoLabel)
		} else {
			icon := "🤡"
			if r.correct != nil && *r.correct {
				icon = "👑"
			}
			line = fmt.Sprintf("%s **#%d** %s %s %s — vote **%s**",
				icon, r.matchID, upper(r.home), score, upper(r.away), upper(r.vote))
		}

		var tags []string
		if r.hasDoubleDown {
			tags = append(tags, "⏫ double-down")
		}
		if r.curseTarget != "" {
			tName := "Unknown"
			if u := users[r.curseTarget]; u != nil {
				tName = u.DisplayName()
			}
			tags = append(tags, fmt.Sprintf("🪄 nguyền **%s**", tName))
		}
		if len(tags) > 0 {
			line += "\n  └ " + strings.Join(tags, " · ")
		}
		lines = append(lines, line)
	}

	totalVoted := 0
	totalCorrect := 0
	totalRandom := 0
	totalAuto := 0
	totalDD := 0
	totalCurses := 0
	for _, r := range history {
		if r.vote != "" {
			totalVoted++
			if r.correct != nil && *r.correct {
				totalCorrect++
			}
		} else {
			if r.hasRandom {
				totalRandom++
			} else {
				totalAuto++
			}
		}
		if r.hasDoubleDown {
			totalDD++
		}
		if r.curseTarget != "" {
			totalCurses++
		}
	}

	winRate := "—"
	if totalVoted > 0 {
		winRate = fmt.Sprintf("%d%%", (totalCorrect*100)/totalVoted)
	}

	summary := fmt.Sprintf("🎯 Tỉ lệ đúng **%s** (%d/%d)", winRate, totalCorrect, totalVoted)
	if totalRandom > 0 {
		summary += fmt.Sprintf(" · 🎲 %d random", totalRandom)
	}
	if totalAuto > 0 {
		summary += fmt.Sprintf(" · 🤖 %d tự động", totalAuto)
	}
	if totalDD > 0 || totalCurses > 0 {
		var parts []string
		if totalDD > 0 {
			parts = append(parts, fmt.Sprintf("⏫ %d double-down", totalDD))
		}
		if totalCurses > 0 {
			parts = append(parts, fmt.Sprintf("🪄 %d curse", totalCurses))
		}
		summary += "\n" + strings.Join(parts, " · ")
	}
	if len(cursedTargetCounts) > 0 {
		type kv struct {
			id    string
			count int
		}
		var kvs []kv
		for id, c := range cursedTargetCounts {
			kvs = append(kvs, kv{id, c})
		}
		sort.Slice(kvs, func(a, b int) bool { return kvs[a].count > kvs[b].count })
		parts := make([]string, 0, len(kvs))
		for _, kv := range kvs {
			name := "Unknown"
			if u := users[kv.id]; u != nil {
				name = u.DisplayName()
			}
			parts = append(parts, fmt.Sprintf("%s (×%d)", name, kv.count))
		}
		summary += "\n🪄 Nguyền: " + strings.Join(parts, ", ")
	}

	desc := fmt.Sprintf("**%s**\n\n%s\n\n%s", tournamentName, joinLines(lines, "\n"), summary)
	if len(desc) > 4096 {
		desc = desc[:4093] + "..."
	}

	embed := discord.Embed{
		Title:       fmt.Sprintf("📜  Lịch sử vote %s", targetDisplayName),
		Description: desc,
		Color:       0x5865F2,
		Footer:      &discord.EmbedFooter{Text: fmt.Sprintf("Hiển thị %d/%d trận gần nhất", len(recent), len(history))},
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
	}
	if targetAvatarURL != "" {
		embed.Thumbnail = &discord.EmbedImage{URL: targetAvatarURL}
	}

	editReply(ctx, b, i, discord.EditMessagePayload{Embeds: []discord.Embed{embed}})
}
